ntoskrnl/mm/ARM3/wslist.cpp

```diff
@@ -0,0 +1,484 @@
+
+        /* Not supported yet */
+        ASSERT(Pfn->u3.e1.PrototypePte == 0);
+        ASSERT(!MI_IS_ROS_PFN(Pfn));
+
+        /* FIXME: Remove this hack when possible */
+        if (Pfn->Wsle.u1.e1.LockedInMemory || (Pfn->Wsle.u1.e1.LockedInWs))
+        {
+            MiReleasePfnLock(OldIrql);
+            continue;
+        }
+
+        /* We can remove it from the list. Save Protection first */
+        ULONG Protection = Entry.u1.e1.Protection;
+        RemoveFromWsList(WsList, Entry.u1.VirtualAddress);
+
+        /* Dirtify the page, if needed */
+        if (PointerPte->u.Hard.Dirty)
+            Pfn->u3.e1.Modified = 1;
+
+        /* Make this a transition PTE */
+        MI_MAKE_TRANSITION_PTE(PointerPte, Page, Protection);
+        KeInvalidateTlbEntry(MiAddressToPte(PointerPte));
+
+        /* Drop the share count. This will take care of putting it in the standby or modified list. */
+        MiDecrementShareCount(Pfn, Page);
+
+        MiReleasePfnLock(OldIrql);
+        Ret++;
+    }
+    return Ret;
+}
+
+/* GLOBAL FUNCTIONS ***********************************************************/
+extern "C"
+{
+
+VOID
+NTAPI
+MiInsertInWorkingSetList(
+    _Inout_ PMMSUPPORT Vm,
+    _In_ PVOID Address,
+    _In_ ULONG Protection)
+{
+    PMMWSL WsList = Vm->VmWorkingSetList;
+
+    /* Make sure that we are holding the right locks. */
+    ASSERT(MM_ANY_WS_LOCK_HELD(PsGetCurrentThread()));
+    MI_ASSERT_PFN_LOCK_HELD();
+
+    PMMPTE PointerPte = MiAddressToPte(Address);
+
+    /* Make sure we are adding a paged-in address */
+    ASSERT(PointerPte->u.Hard.Valid == 1);
+    PMMPFN Pfn1 = MiGetPfnEntry(PFN_FROM_PTE(PointerPte));
+    ASSERT(Pfn1->u3.e1.PageLocation == ActiveAndValid);
+
+    /* Shared pages not supported yet */
+    ASSERT(Pfn1->u1.WsIndex == 0);
+    ASSERT(Pfn1->u3.e1.PrototypePte == 0);
+
+    /* Nor are "ROS PFN" */
+    ASSERT(MI_IS_ROS_PFN(Pfn1) == FALSE);
+
+    Pfn1->u1.WsIndex = GetFreeWsleIndex(WsList);
+    MMWSLENTRY& NewWsle = WsList->Wsle[Pfn1->u1.WsIndex].u1.e1;
+    NewWsle.VirtualPageNumber = reinterpret_cast<ULONG_PTR>(Address) >> PAGE_SHIFT;
+    NewWsle.Protection = Protection;
+    NewWsle.Direct = 1;
+    NewWsle.Hashed = 0;
+    NewWsle.LockedInMemory = 0;
+    NewWsle.LockedInWs = 0;
+    NewWsle.Age = 0;
+    NewWsle.Valid = 1;
+
+    Vm->WorkingSetSize += PAGE_SIZE;
+    if (Vm->WorkingSetSize > Vm->PeakWorkingSetSize)
+        Vm->PeakWorkingSetSize = Vm->WorkingSetSize;
+}
+
+VOID
+NTAPI
+MiRemoveFromWorkingSetList(
+    _Inout_ PMMSUPPORT Vm,
+    _In_ PVOID Address)
+{
+    RemoveFromWsList(Vm->VmWorkingSetList, Address);
+
+    Vm->WorkingSetSize -= PAGE_SIZE;
+}
+
+_Requires_exclusive_lock_held_(WorkingSet->WorkingSetMutex)
+VOID
+NTAPI
+MiInitializeWorkingSetList(_Inout_ PMMSUPPORT WorkingSet)
+{
+    PMMWSL WsList = WorkingSet->VmWorkingSetList;
+
+    /* Initialize some fields */
+    WsList->FirstFree = ULONG_MAX;
+    WsList->Wsle = reinterpret_cast<PMMWSLE>(WsList + 1);
```

Comment:
This member is a pointer to the last entry?

Reply:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ndk/mmtypes.h;hb=0dedb9b47457e86ca57b93856af3da6cb0138a93#l868

The WsList is at the beginning of a page (in hyper space for processes, at MmSystemCacheWorkingSetList for the system one) and the array of MMWSLE starts right after. Then pages get allocated as the array grows.